src/output_queue/synth.py

```diff
@@ -0,0 +1,84 @@
+import math
+from threading import Lock, Thread
+import pyaudio
+import itertools
+import numpy as np
+
+BUFFER_SIZE = 16
+SAMPLE_RATE = 32000
+NOTE_AMPLITUDE = 0.05
+
+SYNTHESIZER_NAME = "CodeToJoy Synthesizer"
+
+class MIDISynthesizer(Thread):
+    """
+    A simple synthesizer that accepts MIDI input from our system and plays it through
+    the system's default audio output device. This mocks a mido output port for interopability.
+
+    This was heavily inspired by this post here:
+    https://python.plainenglish.io/build-your-own-python-synthesizer-part-3-162796b7d351
+    """
+    def __init__(self):
+        Thread.__init__(self)
+
+        self.active = False
+        self.notes_dict = {}
+        self.notes_dict_lock = Lock()
+        self.name = SYNTHESIZER_NAME
+
+    def close(self):
+        """Stops the synthesizer, closes the audio stream, and joins the thread."""
+        self.active = False
+        self.join()
+
+    def send(self, message):
+        """Send a MIDI event to the synthesizer."""
+
+        with self.notes_dict_lock:
+        # Some MIDI streams don't use `note_off` and instead send a `note_on` with velocity 0
+            if (message.type == 'note_off' or message.velocity == 0) and message.note in self.notes_dict:
+                del self.notes_dict[message.note]
+            elif message.type == 'note_on':
+                self.notes_dict[message.note] = self.sound_generator(message.note, message.velocity / 127)
```

Comment:
We divide by 127 to get the amplitude im assuming? Can you throw a comment in for that